docs/dialog.md

```diff
@@ -0,0 +1,51 @@
+* `COMPLETED_PROMPT`
+* `FAILED_PROMPT`
+* `ADVANCED_TO_NEXT_PROMPT`
+* `DRILL_COMPLETED`
+* `NEXT_DRILL_REQUESTED`
+* `OPTED_OUT`
+
+These events are the source of truth for the system, both within the Dialog Context and for other contexts. They are consumed in three places:
+
+* Within the dialog core, the events are used to derive a dialog state object that we persist in DynamoDB.
+* Within the SMS context, we use dialog events (consumed via a DynamoDB stream) to send messages.
+* Within the Drill Progress context, we use dialog events (consumed via a DynamoDB stream) to update a user’s progress across drills.
+
+## Consistency and Ordering
+
+The dialog events are the source of truth for the entire system. If dialog state or drill progress ever become corrupt, we can rebuild them from the dialog events. But we’ll only get accurate results if we process the dialog events in order. So we’ve taken care to make sure that we always process events in order for each user.
+
+How we maintain ordering and consistency:
+
+* **Stream partitioning**
+    * **The command stream is partitioned by phone number**, and each partition has only one consuming lambda. That ensures that we don’t process two commands for one phone number at the same time.
+    * **DynamoDB tables are partitioned by phone number.** We rely on dynamoDB streams to propagate events to other contexts. Each stream partition has only one consuming lambda. That guarantees that each phone number’s events are processed in order.
+* **Event batching**
+    * **All events produced by a single command are persisted together in one “event batch” item in DynamoDB.** Each command can produce multiple events. E.g., `PROMPT_COMPLETED` and `ADVANCED_TO_NEXT_PROMPT` are a common combination. We found that when we persisted events individually, without batching, that we couldn’t guarantee the order that the order the events would appear in the stream.
+    * **Each event batch is tagged with a sequence number.** We obtain the sequence number from the kinesis command stream. Each event batch is tagged with the sequence number of the command that produced the batch. Downstream consumers can track the sequence number to ensure that they don’t update based on old events.
+* **Each command results in one DynamoDB transaction that both updates the dialog state and writes a dialog event batch.** It’s a simple way to ensure that our state and our events are in sync.
+* **Drill content doesn’t change while the user is in the middle of a drill.** When a user starts a drill, we take a snapshot of the drill and store it in dialog state. That snapshot stays in the user’s dialog state until the drill is complete. So modifications to a drill’s content won’t lead to a jarring experience for users who are in the middle of that drill.
+
+## Components
+
+* 1 [lambda](../stopcovid/dialog/aws_lambdas/handle_command.py) that processes commands from the Kinesis stream
+* The Kinesis stream itself
```

Comment:
For consistency, let's refer to the Kinesis stream here as the "dialog command stream"